File: Contests/Codechef_AUG18/Modular GCD/GCDMOD2.cpp
//Problem: https://www.codechef.com/AUG18B/problems/GCDMOD

//CLEARS SUBTASK 1 ONLY - PARTIAL CORRECT

#include <bits/stdc++.h>
#include <cmath>
#include <climits>
#define test int t; cin>>t; while(t--)
#define rep0(n) for(ll i=0;i<(n);i++)
#define rep1(n) for(ll i=1;i<=(n);i++)
using namespace std;

#define MOD 1000000007

typedef long long ll;
ll fastxp(ll base,ll exp,ll mod)
{
  ll res = 1;
  while(exp>0)
  {
    if(exp%2==1)
      res = (res*base)%mod;
    exp/=2;
    base=(base*base)%mod;
  }
  return res;
}


// Stein's Algorithm
ll gcd(ll a, ll b)
{
  if (a == b)
    return a;

  // GCD(0, b) == b; GCD(a, 0) == a,
  // GCD(0, 0) == 0
  if (a == 0)
    return b;
  if (b == 0)
    return a;

  // look for factors of 2
  if (~a & 1) // a is even
  {
    if (b & 1) // b is odd
      return gcd(a >> 1, b);
    else // both a and b are even
      return gcd(a >> 1, b >> 1) << 1;
  }

  if (~b & 1) // a is odd, b is even
    return gcd(a, b >> 1);

  // reduce larger number
  if (a > b)
    return gcd((a - b) >> 1, b);

  return gcd((b - a) >> 1, a);
}


int main()
{
    /*ios_base::sync_with_stdio(false); 
    cin.tie(NULL); 
    cout.tie(NULL);*/
    test
    {
        ll a,b,n;
      	cin>>a>>b>>n;
      	ll two = (abs(a-b)) ;
      	ll one = (fastxp(a,n,MOD) + fastxp(b,n,MOD)) ;
      	//cout<<one<<" "<<two<<endl; //DEBUG LINE
      	cout << gcd(one,two)%MOD << endl;
    }
    return 0;
}

File: Contests/Codechef_AUG18/Modular GCD/GCDMOD2.py
#Problem: https://www.codechef.com/AUG18B/problems/GCDMOD
#Author : Amitrajit Bose

MOD = (10**9)+7
# Function to return gcd of a and b
def gcd(a, b): 
    if a == 0 :
        return b 
    
    return gcd(b%a, a)

def power(x, y, p) :
    res = 1
    x = x % p 
 
    while (y > 0) :
        if ((y & 1) == 1) :
            res = (res * x) % p
 
        # y must be even now
        y = y >> 1      # y = y/2
        x = (x * x) % p
         
    return res

# MAIN
test=int(input())
for _ in range(test):
    a,b,n = [int(x) for x in input().strip().split()]
    one = (power(a, n, MOD) + power(b, n, MOD))
    two = abs(a-b)
    out = gcd(one, two) % MOD
    print(out)